### ISSUE TITLE
Graph::ApplyGradients ResetCaching performance improvement

### ISSUE THREAD

**Issue by evsmithx** [fetchai/ledger#1554]
When two nodes share a Trainable op ptr, both nodes should reset their cache when the op is trained. But the graph only knows about the first node and not the copied one. The current fix for this is to reset all the node caches when the nodes are trained. A better fix would be to make Graph.trainable_nodes_  contain only nodes which are in Graph.nodes_ and make GetTrainables a recursive call.

**Comment by baykaner** [comment]
this problem is now slightly different. Since trainable_nodes_ and trainable_lookup_ in graph have been reworked, it is now already the case that graph.trainable_lookup_ is recursive, so this aspect is no longer problematic.

However, shared layers ops' do not share gradient accumulations. and when these gradient accumulations are applied, layers are not updated in sync.

**Comment by baykaner** [comment]
To implement this suggested caching improvement, the following would have to be done in graph:
1. collect a list of nodes (not just trainables) touched by any backpropagate call
2. apply gradients and reset only those caches
3. reset all shared layers/ops (because these may be updated despite not being backpropagated directly)

This will improve forward propagation time for networks that have a topology similar to the below diagram. In this example, repeated training on Loss function 1 does not require resetting the cache for layers that occur before the shared layer (e.g. layer 2):

![shared_layer_reset_cache](https://user-images.githubusercontent.com/40827646/66479127-c05e2c00-ea93-11e9-8749-76e6f4559730.png)
